File: myFinance_improved_V2/lib/features/cash_ending/data/repositories/currency_repository_impl.dart
// lib/features/cash_ending/data/repositories/currency_repository_impl.dart

import '../../domain/entities/currency.dart';
import '../../domain/entities/denomination.dart';
import '../../domain/repositories/currency_repository.dart';
import '../datasources/currency_remote_datasource.dart';
import '../models/freezed/currency_dto.dart';
import '../models/freezed/denomination_dto.dart';
import 'base_repository.dart';

/// Repository Implementation for Currencies (Data Layer)
///
/// Implements the domain repository interface.
/// Coordinates between datasource (Supabase) and domain entities.
/// Handles data transformation and error mapping using BaseRepository.
class CurrencyRepositoryImpl extends BaseRepository
    implements CurrencyRepository {
  final CurrencyRemoteDataSource _remoteDataSource;

  CurrencyRepositoryImpl({
    CurrencyRemoteDataSource? remoteDataSource,
  }) : _remoteDataSource = remoteDataSource ?? CurrencyRemoteDataSource();

  @override
  Future<List<Currency>> getCompanyCurrenciesWithExchangeRates({
    required String companyId,
    DateTime? rateDate,
  }) async {
    return executeWithErrorHandling(
      () async {
        if (companyId.isEmpty) {
          return [];
        }

        // RPC call (1 network request for all data)
        final data = await _remoteDataSource.getCompanyCurrenciesWithExchangeRates(
          companyId: companyId,
          rateDate: rateDate,
        );

        // JSON → DTO → Entity conversion
        // Use fromRpcJson to inject currency_id into denominations
        return data
            .map((json) => CurrencyDto.fromRpcJson(json).toEntity())
            .toList();
      },
      operationName: 'getCompanyCurrenciesWithExchangeRates',
    );
  }

  @override
  Future<List<Denomination>> getDenominationsByCurrency({
    required String companyId,
    required String currencyId,
  }) async {
    return executeWithErrorHandling(
      () async {
        if (companyId.isEmpty || currencyId.isEmpty) {
          return [];
        }

        // Call remote datasource
        final data = await _remoteDataSource.getDenominationsByCurrency(
          companyId: companyId,
          currencyId: currencyId,
        );

        // Convert JSON to DTOs then to entities
        return data
            .map((json) => DenominationDto.fromJson(json).toEntity())
            .toList();
      },
      operationName: 'getDenominationsByCurrency',
    );
  }
}

File: website/src/features/currency/data/repositories/CurrencyRepositoryImpl.ts
import { ICurrencyRepository, CurrencyResult, CurrencyTypesResult } from '../../domain/repositories/ICurrencyRepository';
import { CurrencyDataSource } from '../datasources/CurrencyDataSource';
import { CurrencyModel } from '../models/CurrencyModel';

export class CurrencyRepositoryImpl implements ICurrencyRepository {
  private dataSource = new CurrencyDataSource();
  async getCurrencies(companyId: string): Promise<CurrencyResult> {
    try {
      const dtos = await this.dataSource.getCurrencies(companyId);
      const currencies = dtos.map(dto => CurrencyModel.toEntity(dto));
      return { success: true, data: currencies };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed' };
    }
  }
  async getAllCurrencyTypes(): Promise<CurrencyTypesResult> {
    try {
      const data = await this.dataSource.getAllCurrencyTypes();
      return { success: true, data };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed' };
    }
  }
  async setDefaultCurrency(currencyId: string, companyId: string) {
    try {
      await this.dataSource.setDefaultCurrency(currencyId, companyId);
      return { success: true };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed' };
    }
  }
  async updateExchangeRate(currencyId: string, companyId: string, newRate: number, userId: string) {
    try {
      await this.dataSource.updateExchangeRate(currencyId, companyId, newRate, userId);
      return { success: true };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed' };
    }
  }
  async addCurrency(currencyId: string, companyId: string, exchangeRate: number, userId: string) {
    try {
      await this.dataSource.addCurrency(currencyId, companyId, exchangeRate, userId);
      return { success: true };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed' };
    }
  }
  async removeCurrency(currencyId: string, companyId: string) {
    try {
      await this.dataSource.removeCurrency(currencyId, companyId);
      return { success: true };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Failed' };
    }
  }
}
